e2e/consul/script_checks.go

```diff
@@ -0,0 +1,212 @@
+package consul
+
+import (
+	"time"
+
+	capi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/nomad/e2e/e2eutil"
+	"github.com/hashicorp/nomad/e2e/framework"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+type ScriptChecksE2ETest struct {
+	framework.TC
+	jobIds []string
+}
+
+func init() {
+	framework.AddSuites(&framework.TestSuite{
+		Component:   "ScriptChecks",
+		CanRunLocal: true,
+		Consul:      true,
+		Cases: []framework.TestCase{
+			new(ScriptChecksE2ETest),
+		},
+	})
+}
+
+func (tc *ScriptChecksE2ETest) BeforeAll(f *framework.F) {
+	// Ensure cluster has leader before running tests
+	e2eutil.WaitForLeader(f.T(), tc.Nomad())
+	// Ensure that we have at least 1 client node in ready state
+	e2eutil.WaitForNodesReady(f.T(), tc.Nomad(), 1)
+}
+
+// serviceStatus asserts the health of the service and returns the
+// []ServiceEntry for further checking
+func serviceStatus(require *require.Assertions,
+	consulClient *capi.Client, serviceName, expectedStatus string) []*capi.ServiceEntry {
+	services, _, err := consulClient.Health().Service(serviceName, "", false, nil)
+	require.NoError(err, "expected no error for %q, got %v", serviceName, err)
+	for _, check := range services[0].Checks {
+		if check.CheckID != "serfHealth" {
+			require.Equal(expectedStatus, check.Status,
+				"expected check %s for %s to be %q, got %q",
+				check.CheckID, check.ServiceName, expectedStatus, check.Status,
+			)
+		}
+	}
+	return services
+}
+
+// serviceRegistered asserts that the service has been deregistered from Consul
+func serviceDeregistered(require *require.Assertions,
+	consulClient *capi.Client, serviceName string) {
+	services, _, err := consulClient.Health().Service(serviceName, "", false, nil)
+	require.NoError(err, "expected no error for %q, got %v", serviceName, err)
+	require.Len(services, 0)
+}
+
+// TestTaskScriptCheck runs a job with a single task group with several services
+// and associated script checks. It updates, stops, etc. the job to verify
+// that script checks are re-registered as expected.
+func (tc *ScriptChecksE2ETest) TestGroupScriptCheck(f *framework.F) {
+	nomadClient := tc.Nomad()
+	uuid := uuid.Generate()
+	require := require.New(f.T())
+	consulClient := tc.Consul()
+
+	jobId := "checks_group" + uuid[0:8]
+	tc.jobIds = append(tc.jobIds, jobId)
+
+	// Job run: verify that checks were registered in Consul
+	allocs := e2eutil.RegisterAndWaitForAllocs(f.T(),
+		nomadClient, "consul/input/checks_group.nomad", jobId)
+	require.Equal(1, len(allocs))
+	require.Eventually(func() bool {
+		serviceStatus(require, consulClient, "group-check-1", capi.HealthPassing)
+		// TODO(tgross): we want to verify this goes from unhealthy to
```

Comment:
I think I've figured out what I'll do here: for the `Crash a task: verify that checks become unhealthy and then healthy` test, I need to reach into the task (probably via nomad exec). So I'll update one of these checks to read a non-existent file in the task, verify Consul sees it as unhealthy, nomad exec into the task to touch the file, and that'll cause the health check to flip to healthy.